### PR TITLE
[Dubbo-6720] fix bug same interface unexport and export fail. also support hotload service

### PR DESCRIPTION
**Pull request by owen200008**
## What is the purpose of the change
fix bug same interface unexport and export fail.
delegate exportmap and support hotload export

## Brief changelog
1. use interface to oper exportmap 
2. remove export must make sure the same export

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by AlbumenJ** [comment]
@owen200008 Hi, thanks for your contribution.

Please merge the latest master branch to resolve confilcts.

**Comment by owen200008** [comment]
@AlbumenJ I have resolve confilcts
please check the result.

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractProxyProtocol.java [hunk 1]
```diff
@@ -1,281 +1,279 @@
...
-
-        @Override
-        public boolean isClosed() {
-            return false;
-        }
-    }
-
-
-}
+/*
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractProxyProtocol.java:1]: please recover file endline format of this file

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/DelegateExporterMap.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/DelegateExporterMap.java:1]: also for this file

File: dubbo-rpc/dubbo-rpc-injvm/src/main/java/org/apache/dubbo/rpc/protocol/injvm/InjvmExporter.java [hunk 3]
```diff
@@ -1,46 +1,42 @@
...
-        exporterMap.put(key, this);
-    }
-
-    @Override
-    public void afterUnExport() {
-        exporterMap.remove(key);
-    }
-
-}
+/*
```
Comment (AlbumenJ) [inline dubbo-rpc/dubbo-rpc-injvm/src/main/java/org/apache/dubbo/rpc/protocol/injvm/InjvmExporter.java:1]: also

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Comment by AlbumenJ** [comment]
image

Please fix codestyle check.

**Comment by owen200008** [comment]
@AlbumenJ finish thanks!

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractProtocol.java [hunk 4]
```diff
@@ -42,7 +43,7 @@
 /**
  * abstract ProtocolSupport.
  */
-public abstract class AbstractProtocol implements Protocol {
+public abstract class AbstractProtocol implements DelegateExporterMap, Protocol {
```
Comment (chickenlj) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractProtocol.java:46]: I think it's better to make `DelegateExporterMap` as AbstractProtocol's property than as its super interface.

That is, AbstractProtocol has a `DelegateExporterMap` rather is a `DelegateExporterMap`

**Review by chickenlj — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractProtocol.java [hunk 5]
```diff
@@ -1,116 +1,148 @@
...
+/**
+ * abstract ProtocolSupport.
+ */
+public abstract class AbstractProtocol implements Protocol {
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    protected final Map<String, Exporter<?>> exporterMap = new ConcurrentHashMap<String, Exporter<?>>();
+
+    protected final DelegateExporterMap delegateExporterMap = new DelegateExporterMap() {
```
Comment (chickenlj) [inline dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractProtocol.java:52]: Can we merge `exporterMap` and `delegateExporterMap` by defining DelegateExporterMap as a Map or have it referencing a map?

**Review by chickenlj — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractProtocol.java (hunk 5 above)
Comment (chickenlj) [reply dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractProtocol.java:52]: Pls., also check the failed unit tests. We'll try to merge it quickly this time once ready, thanks.

**Comment by AlbumenJ** [comment]
Great! @owen200008  Would you please pick these changes to 3.0?

**Pull request merged**